File: py-server/app/controllers/image_controller.py
import os
from flask import request, jsonify
from app.models.album import Album
from app.models.image import Image
from app.helpers.s3_helper import upload_file
from app.helpers.rekognition_helper import get_labels, extract_text
from app.helpers.translate_helper import translate_text
from config.db import db


def get_images():
    try:
        images = Image.query.all()
        return jsonify([image.to_dict() for image in images]), 200
    except Exception as e:
        return jsonify({'message': f'Error interno del servidor: {str(e)}'}), 500


def get_image(image_id):
    try:
        image = Image.query.get(image_id)
        if image:
            return jsonify(image.to_dict()), 200
        else:
            return jsonify({'message': 'Imagen no encontrada.'}), 404
    except Exception as e:
        return jsonify({'message': f'Error interno del servidor: {str(e)}'}), 500


def get_image_labels(image_id):
    try:
        image = Image.query.get(image_id)
        if not image:
            return jsonify({'message': 'Imagen no encontrada.'}), 404
        
        # Obtener el nombre del bucket de la variable de entorno
        s3_bucket_name = os.getenv('AWS_BUCKET_NAME')

        # Extraer el nombre del objeto S3 desde la URL de la imagen
        s3_object_name = image.url.split('/', 3)[-1]
        
        # Obtener etiquetas de la imagen en S3
        labels = get_labels(s3_bucket_name, s3_object_name)
        
        return jsonify({'labels': labels}), 200
    except Exception as e:
        return jsonify({'message': f'Error interno del servidor: {str(e)}'}), 500


def get_image_desc_translations(image_id):
    try:
        image = Image.query.get(image_id)
        if not image:
            return jsonify({'message': 'Imagen no encontrada.'}), 404

        # Obtener la descripción de la imagen
        description = image.description

         # Obtener traducciones de la descripción
        translations = translate_text(description)
        
        return jsonify({'translations': translations}), 200
    except Exception as e:
        return jsonify({'message': f'Error interno del servidor: {str(e)}'}), 500


def add_image():
    try:
        data = request.form
        name = data.get('name')
        description = data.get('description')
        url = request.files.get('url')
        album_id = data.get('album_id')

        # Validar la entrada
        if not all([name, description, url, album_id]):
            return jsonify({'message': 'Todos los campos son necesarios.'}), 400

        # Obtener el álbum
        album = Album.query.get(album_id)
        if not album:
            return jsonify({'message': 'Álbum no encontrado.'}), 404
        
        # Verificar si se pueden subir imágenes al álbum
        # Buscar el primer álbum con el mismo user_id
        first_album_with_user_id = Album.query.filter_by(user_id=album.user_id).order_by(Album.album_id).first()
        if first_album_with_user_id and first_album_with_user_id.album_id == int(album_id):
            return jsonify({'message': f'El álbum "{album.name}" no permite subir imágenes.'}), 403
        
        # Subir la imagen al bucket S3
        prefix = f'Fotos_Publicadas/Usuario-{album.user_id}/Album-{album_id}'
        image_url = upload_file(url, f'{prefix}/{url.filename}')

        # Crear y agregar la nueva imagen a la base de datos
        new_image = Image(
            name=name,
            description=description,
            url=image_url,
            album_id=album_id
        )
        db.session.add(new_image)
        db.session.commit()

        return jsonify(new_image.to_dict()), 201
    except Exception as e:
        return jsonify({'message': f'Error interno del servidor: {str(e)}'}), 500


def extract_text_from_image():
    try:
        image = request.files.get('image')

        # Verificar si la imagen fue proporcionada
        if not image:
            return jsonify({'message': 'Se requiere la imagen.'}), 400
        
        # Obtener texto de la imagen
        text = extract_text(image)
        
        return jsonify({'text': text}), 200
    except Exception as e:
        return jsonify({'message': f'Error interno del servidor: {str(e)}'}), 500

File: js-server/controllers/image_controller.js
const express = require('express');
const bcrypt = require('bcrypt');
const multer = require('multer');
const AWS = require('aws-sdk');
const dotenv = require('dotenv');
const conexion  = require('../conexion/conexion')
const deleteFiles = require('../eliminarS3/eliminar')

dotenv.config();

//Configuracion de las credenciales para AWS
AWS.config.update({
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
    region: process.env.AWS_REGION,
    //bucketName: process.env.AWS_BUCKET_NAME
});

const s3 = new AWS.S3();
//const upload = multer();
//const router = express.Router();

const nameBucket = process.env.AWS_BUCKET_NAME //nombre del bucket
const region = process.env.AWS_BUCKET_REGION //region

//Funcion para subir archivos a S3
const uploadFile =  (file, filePath) =>{
    const parametros = {
        Bucket: process.env.AWS_BUCKET_NAME,
        Key: filePath,
        Body: file.buffer,
        ContentType: file.mimetype
    };
    return s3.upload(parametros).promise().then(data => data.Location).catch(err =>{
        throw new Error(`Error al subir archivo a S3: ${err.message}`);
    });
    
};


//Obtener imagen por ID
const get_image = async (req, res) => {
    try {
        const imageId = req.params.image_id;

        // Consulta a la tabla Imagen para obtener el id del album
        const queryImage = 'SELECT * FROM Imagen WHERE id_imagen = ?';
        const [imageResults] = await conexion.promise().query(queryImage, [imageId]);

        if (imageResults.length === 0) {
            return res.status(404).json({ message: 'Imagen no encontrada' });
        }

        const image = imageResults[0];
        const albumId = image.id_album;

        // Consulta a la tabla Album para obtener el id del usuario
        const queryAlbum = 'SELECT * FROM Album WHERE id_album = ?';
        const [albumResults] = await conexion.promise().query(queryAlbum, [albumId]);

        if (albumResults.length === 0) {
            return res.status(404).json({ message: 'Álbum no encontrado' });
        }

        const album = albumResults[0];
        const userId = album.id_usuario;

        // Consulta a la tabla Usuario para obtener los detalles del usuario
        const queryUser = 'SELECT * FROM Usuario WHERE id_usuario = ?';
        const [userResults] = await conexion.promise().query(queryUser, [userId]);

        if (userResults.length === 0) {
            return res.status(404).json({ message: 'Usuario no encontrado' });
        }

        const user = userResults[0];

        // respuesta
        const response = {
            image_id: image.id_imagen,
            name: image.nombre,
            description: image.descripcion,
            url: image.url,
            album: {
                album_id: album.id_album,
                name: album.nombre,
                user: {
                    user_id: user.id_usuario,
                    username: user.nombre_usuario,
                    email: user.correo,
                    image: user.imagen,
                    login_image: user.login_imagen,
                    image_key: user.imagen_clave
                }
            }
        };

        res.status(200).json(response);

    } catch (err) {
        res.status(500).json({ message: `Error en la consulta: ${err.message}` });
    }
};


//Obtener todas las imagenes
const get_images = (req, res) => {
    const queryImages = 'SELECT * FROM Imagen';
    
    conexion.query(queryImages, (err, images) => {
        if (err) {
            return res.status(500).json({ message: `Error interno del servidor: ${err.message}` });
        }

        if (images.length === 0) {
            return res.status(404).json({ message: 'No se encontraron imágenes' });
        }

        // Obtener todos los IDs de álbumes
        const albumIds = [...new Set(images.map(image => image.id_album))];
        
        // Consultar álbumes
        const queryAlbums = 'SELECT * FROM Album WHERE id_album IN (?)';
        conexion.query(queryAlbums, [albumIds], (err, albums) => {
            if (err) {
                return res.status(500).json({ message: `Error en la consulta de álbumes: ${err.message}` });
            }

            // Obtener todos los IDs de usuarios
            const userIds = [...new Set(albums.map(album => album.id_usuario))];
            
            // Consultar usuarios
            const queryUsers = 'SELECT * FROM Usuario WHERE id_usuario IN (?)';
            conexion.query(queryUsers, [userIds], (err, users) => {
                if (err) {
                    return res.status(500).json({ message: `Error en la consulta de usuarios: ${err.message}` });
                }

                // Mapear datos para acceso rápido
                const albumMap = new Map(albums.map(album => [album.id_album, album]));
                const userMap = new Map(users.map(user => [user.id_usuario, user]));

                // Combinar la información
                const response = images.map(image => {
                    const album = albumMap.get(image.id_album);
                    const user = userMap.get(album.id_usuario);

                    return {
                        image_id: image.id_imagen,
                        name: image.nombre,
                        description: image.descripcion,
                        url: image.url,
                        album: {
                            album_id: album.id_album,
                            name: album.nombre,
                            user: {
                                user_id: user.id_usuario,
                                username: user.nombre_usuario,
                                email: user.correo,
                                image: user.imagen,
                                login_image: user.login_imagen,
                                image_key: user.imagen_clave
                            }
                        }
                    };
                });

                res.status(200).json(response);
            });
        });
    });
};


//Agregar imagenes
const add_image = async (req, res) => {
    try {
        const { name, description, album_id } = req.body;

        // Verificar que todos los campos
        if (!name || !description || !album_id || !req.file) {
            return res.status(400).json({ message: 'Todos los campos son necesarios.' });
        }

        const albumId = parseInt(album_id, 10);

        // Obtener el álbum
        const getAlbumQuery = 'SELECT * FROM Album WHERE id_album = ?';
        const [albumResults] = await conexion.promise().query(getAlbumQuery, [albumId]);

        if (!albumResults || albumResults.length === 0) {
            return res.status(404).json({ message: 'Álbum no encontrado.' });
        }

        const album = albumResults[0];

        // Verificar si se pueden subir imágenes al álbum
        const getFirstAlbumQuery = 'SELECT * FROM Album WHERE id_usuario = ? ORDER BY id_album LIMIT 1';
        const [firstAlbumResults] = await conexion.promise().query(getFirstAlbumQuery, [album.id_usuario]);

        if (!firstAlbumResults || firstAlbumResults.length === 0) {
            return res.status(404).json({ message: 'No se encontraron álbumes para este usuario.' });
        }

        if (firstAlbumResults[0].id_album === albumId) {
            return res.status(403).json({ message: `El álbum "${album.nombre}" no permite subir imágenes.` });
        }

        // Subir la imagen al bucket S3
        const prefix = `Fotos_Publicadas/Usuario-${album.id_usuario}/Album-${albumId}`;
        const imageFile = req.file;
        const filePath = `${prefix}/${imageFile.originalname}`;
        const imageUrl = `https://${nameBucket}.s3.${region}.amazonaws.com/${filePath}`;

        await uploadFile(imageFile, filePath);

        // Insertar la imagen en la base de datos
        const addImageQuery = 'INSERT INTO Imagen (nombre, descripcion, url, id_album) VALUES (?, ?, ?, ?)';
        const [insertResult] = await conexion.promise().query(addImageQuery, [name, description, imageUrl, albumId]);

        if (!insertResult) {
            return res.status(500).json({ message: 'Error al agregar la imagen.' });
        }

        // Obtener la nueva imagen para respuesta
        const newImageId = insertResult.insertId;
        const getImageQuery = 'SELECT * FROM Imagen WHERE id_imagen = ?';
        const [newImageResults] = await conexion.promise().query(getImageQuery, [newImageId]);

        if (!newImageResults || newImageResults.length === 0) {
            return res.status(404).json({ message: 'Imagen no encontrada después de la inserción.' });
        }

        const newImage = newImageResults[0];

        // Obtener información del álbum y del usuario relacionado
        const getAlbumUserQuery = `
            SELECT 
                a.id_album, a.nombre AS album_name, u.id_usuario AS user_id, u.nombre_usuario, u.correo, u.imagen AS image, u.login_imagen, u.imagen_clave 
            FROM Album a 
            JOIN Usuario u ON a.id_usuario = u.id_usuario 
            WHERE a.id_album = ?`;
        const [albumUserInfo] = await conexion.promise().query(getAlbumUserQuery, [albumId]);

        if (!albumUserInfo || albumUserInfo.length === 0) {
            return res.status(404).json({ message: 'Álbum o usuario no encontrado.' });
        }

        const albumInfo = albumUserInfo[0];
        //console.log(albumInfo)

        // Responder con la nueva imagen agregada y detalles del álbum y usuario
        return res.status(201).json({
            image_id: newImage.id_imagen,
            name: newImage.nombre,
            description: newImage.descripcion,
            url: newImage.url,
            album: {
                album_id: albumInfo.id_album,
                name: albumInfo.album_name,
                user: {
                    user_id: albumInfo.user_id,
                    username: albumInfo.nombre_usuario,
                    email: albumInfo.correo,
                    image: albumInfo.image, //? `https://${nameBucket}.s3.amazonaws.com/${albumInfo.image}` : null,
                    login_image: albumInfo.login_imagen,
                    image_key: albumInfo.imagen_clave
                }
            }
        });
    } catch (err) {
        return res.status(500).json({ message: `Error interno del servidor: ${err.message}` });
    }
};




module.exports = {add_image,get_images,get_image}